sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,564 @@
+        for sum_fn, w in _steps_data:
+            sum_fn(v, w)
+            v = w
+        return v
+
+    return sum_reduction
+
+
+@cached_kernel_factory
+def make_sum_reduction_kernel_2(n, dim, local_size):
+    global_size = math.ceil(n / local_size) * local_size
+    f32zero = float32(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        thread = dpex.get_global_id(0)
+        if thread >= n:
+            return
+        tmp = f32zero
+        for d in range(dim):
+            tmp += v[d, thread]
+        w[thread] = tmp
+
+    return sum_reduction_kernel[global_size, local_size]
+
+
+@cached_kernel_factory
+def make_sum_reduction_kernel_3(dim0, dim1, dim2, local_size):
+    nb_groups_for_dim2 = math.ceil(dim2 / local_size)
+    nb_threads_for_dim2 = nb_groups_for_dim2 * local_size
+    global_size = nb_threads_for_dim2 * dim1
+    f32zero = float32(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group = dpex.get_group_id(0)
+        thread = dpex.get_local_id(0)
+        d1 = group // nb_groups_for_dim2
+        d2 = ((group % nb_groups_for_dim2) * local_size) + thread
+        if d2 >= dim2:
+            return
+        tmp = f32zero
+        for d in range(dim0):
+            tmp += v[d, d1, d2]
+        w[d1, d2] = tmp
+
+    return sum_reduction_kernel[global_size, local_size]
+
+
+@cached_kernel_factory
+def make_fused_kernel_fixed_window(
+    n,
+    dim,
+    n_clusters,
+    warp_size,
+    l2_cache_size,
+    window_length_multiple,
+    cluster_window_per_thread_group,
+    number_of_load_iter,
+):
+    r = warp_size * window_length_multiple
+    local_size = r * cluster_window_per_thread_group
+    h = number_of_load_iter * cluster_window_per_thread_group
+
+    n_cluster_groups = math.ceil(n_clusters / r)
+    global_size = (math.ceil(n / local_size)) * (local_size)
+    n_dim_windows = math.ceil(dim / h)
+    window_shm_shape = (h, r)
+
+    inf = float32(math.inf)
+    f32zero = float32(0.0)
+    f32one = float32(1.0)
+
+    nb_cluster_items = n_clusters * (dim + 1)
+    nb_cluster_bytes = 4 * nb_cluster_items
+    l2_cache_size_allocation = 0.9
+    nb_centroids_private_copies = int(
+        (l2_cache_size * l2_cache_size_allocation) // nb_cluster_bytes
+    )
+
+    @dpex.kernel
+    def fused_kernel_fixed_window(
+        X,
+        current_centroids,
+        centroids_half_l2_norm,
+        inertia,
+        centroids_private_copies,
+        centroid_counts_private_copies,
+    ):
+        global_thread = dpex.get_global_id(0)
+        local_thread = dpex.get_local_id(0)
+
+        window_shm = dpex.local.array(shape=window_shm_shape, dtype=float32)
+        local_centroids_half_l2_norm = dpex.local.array(shape=r, dtype=float32)
+        partial_scores = dpex.private.array(shape=r, dtype=float32)
+
+        first_centroid_global_idx = 0
+
+        min_idx = 0
+        min_score = inf
+
+        window_col = local_thread % r
+        window_row_offset = local_thread // r
```

Comment:
I think it would be more natural to invoke this kernel using a 2D grid instead of doing this and using offsets induced by `r`.

```suggestion
        window_col = local_thread % r
        window_row_offset = local_thread // r
```